torch/_inductor/runtime/proton_utils.py

```diff
@@ -0,0 +1,325 @@
+"""
+Utilities for post-processing Triton Proton profiling trace files.
+
+This module provides functions to transform proton-generated Chrome trace files
+for better visualization in Perfetto/Chrome tracing tools.
+"""
+
+import gzip
+import json
+import os
+import re
+from typing import Any
+
+from torch.utils._ordered_set import OrderedSet
+
+
+def _write_trace(path: str, data: dict[str, Any]) -> None:
+    """Write trace data to a gzipped JSON file."""
+    with gzip.open(path, "wt", encoding="utf-8") as f:
+        json.dump(data, f)
+
+
+def _read_trace(path: str) -> dict[str, Any]:
+    """Read trace data from a JSON file (gzipped or not)."""
+    if path.endswith(".gz"):
+        with gzip.open(path, "rt", encoding="utf-8") as f:
+            return json.load(f)
+    else:
+        with open(path) as f:
+            return json.load(f)
+
+
+def _get_base_name(trace_path: str) -> str:
+    """Get the base name of a trace file, stripping known extensions."""
+    base_name = os.path.basename(trace_path)
+    for ext in [".trace.json.gz", ".chrome_trace", ".json.gz", ".json"]:
+        if base_name.endswith(ext):
+            return base_name[: -len(ext)]
+    return os.path.splitext(base_name)[0]
+
+
+def _group_events_by_sm(events: list[dict[str, Any]]) -> list[dict[str, Any]]:
+    """
+    Group events so all CTAs on the same SM share one track.
+
+    Transforms:
+    - pid: "Core0 CTA6" -> "Core0"
+    - tid: "warp0" -> "CTA6 warp0"
+    """
+    core_cta_pattern = re.compile(r"^(.*?)\s*(Core\d+)\s+(CTA\d+)$")
+
+    for event in events:
+        pid = event.get("pid", "")
+        tid = event.get("tid", "")
+
+        match = core_cta_pattern.match(str(pid))
+        if match:
+            prefix = match.group(1).strip()
+            core = match.group(2)
+            cta = match.group(3)
+            event["pid"] = f"{prefix} {core}" if prefix else core
+            event["tid"] = f"{cta} {tid}" if tid else cta
+
+    return events
+
+
+def _group_events_per_cta_occupancy(
```

Comment:
just curious - does proton / triton have any facilities for processing / visualization ?